File: clients/go/delivery.go
package fibril

// The delivery path: subscribe, receive pushed messages on a per-subscription
// channel, and settle them with complete/fail/retry. The run goroutine owns the
// sub-id -> channel map and pushes deliveries; the channel is buffered to the effective
// prefetch, so the broker (which never sends beyond prefetch unacked) cannot
// outrun it and the run goroutine never blocks.

import (
	"context"
	"time"
)

// Delivery is one message pushed to a subscription.
type Delivery struct {
	Topic           string
	Group           *string
	Partition       uint32
	Payload         []byte
	ContentType     ContentType
	Headers         Headers
	DeliveryTag     DeliveryTag
	Offset          uint64
	Published       uint64
	PublishReceived uint64
	SubID           uint64
	// AutoAck is true when the broker already settled this delivery server-side
	// (an auto-ack subscription). Settling is then unnecessary.
	AutoAck bool

	reqID uint64 // the DELIVER frame's request id, reused when settling
	// settle routes settlement to whatever engine is currently live for the
	// connection (keyed by incarnation), not the engine this arrived on, so an ack
	// after a reconnect reaches the current engine or reports a StaleDeliveryError.
	settle      *settleContext
	incarnation uint64
}

// Complete settles this delivery as processed. It routes to whatever engine is
// currently live for the connection, so it stays correct across a reconnect (and
// when deliveries from several partitions are fanned in). A non-resumed reconnect
// makes it return a *StaleDeliveryError. Unnecessary for an auto-ack delivery, but
// harmless.
func (d Delivery) Complete() error {
	e, err := d.settle.currentOrStale(d.incarnation)
	if err != nil {
		return err
	}
	return e.Complete(d)
}

// Retry requeues this delivery immediately for redelivery.
func (d Delivery) Retry() error {
	e, err := d.settle.currentOrStale(d.incarnation)
	if err != nil {
		return err
	}
	return e.Nack(d, true, nil)
}

// Fail settles this delivery as a terminal failure: it is not requeued, so it is
// dead-lettered or dropped per the queue's policy.
func (d Delivery) Fail() error {
	e, err := d.settle.currentOrStale(d.incarnation)
	if err != nil {
		return err
	}
	return e.Nack(d, false, nil)
}

// RetryAfter requeues this delivery for redelivery no sooner than delay from now.
func (d Delivery) RetryAfter(delay time.Duration) error {
	e, err := d.settle.currentOrStale(d.incarnation)
	if err != nil {
		return err
	}
	notBefore := time.Now().UnixMilli() + delay.Milliseconds()
	if notBefore < 0 {
		notBefore = 0
	}
	nb := uint64(notBefore)
	return e.Nack(d, true, &nb)
}

// Subscription is a live single-partition subscription. Deliveries yields
// messages until the connection closes (the channel is then closed); settle each
// with its Complete/Fail/Retry.
type Subscription struct {
	SubID      uint64
	Topic      string
	Partition  uint32
	Group      *string
	MemberID   *UUID
	Deliveries <-chan Delivery

	reason *closeReasonCell
	engine *Engine
}

// CloseReason returns why the subscription ended, once its Deliveries channel
// has closed. It is nil while the subscription is live or after a clean local
// close, and set to the typed reason (topic deleted, owner moved, a reconcile
// verdict, ...) when the broker or a reconcile ended it.
func (s *Subscription) CloseReason() *CloseReason {
	if s.reason == nil {
		return nil
	}
	return s.reason.get()
}

// Subscribe opens a subscription and returns it once the broker confirms. The
// returned Subscription's Deliveries channel receives pushed messages. A plain
// subscribe is remembered for reconnect reconcile.
func (e *Engine) Subscribe(ctx context.Context, req Subscribe) (*Subscription, error) {
	return e.subscribe(ctx, req, false)
}

// subscribeSupervised opens a subscription that is not remembered for reconcile,
// since its supervisor re-subscribes on a drop instead.
func (e *Engine) subscribeSupervised(ctx context.Context, req Subscribe) (*Subscription, error) {
	return e.subscribe(ctx, req, true)
}

func (e *Engine) subscribe(ctx context.Context, req Subscribe, noReconcile bool) (*Subscription, error) {
	sr := make(chan subResult, 1)
	reqCopy := req
	select {
	case e.cmdCh <- command{op: opSubscribe, body: encodeSubscribe(req), subReply: sr, autoAck: req.AutoAck, noReconcile: noReconcile, sub: &reqCopy}:
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-e.done:
		return nil, e.err()
	}
	select {
	case r := <-sr:
		return r.sub, r.err
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-e.done:
		return nil, e.err()
	}
}

// SubscribeStream opens a Plexus (fan-out stream) subscription of one partition.
// The broker confirms it with the shared SUBSCRIBE_OK, so it delivers on a
// per-subscription channel like a queue subscription.
func (e *Engine) SubscribeStream(ctx context.Context, req SubscribeStream) (*Subscription, error) {
	sr := make(chan subResult, 1)
	// Streams resume by durable cursor, not the reconcile registry.
	select {
	case e.cmdCh <- command{op: opSubscribeStream, body: encodeSubscribeStream(req), subReply: sr, autoAck: req.AutoAck, noReconcile: true}:
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-e.done:
		return nil, e.err()
	}
	select {
	case r := <-sr:
		return r.sub, r.err
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-e.done:
		return nil, e.err()
	}
}

// Complete settles a delivery as processed. It is a no-op-worthy call for an
// auto-ack delivery (already settled server-side) but harmless. Fire-and-forget.
func (e *Engine) Complete(d Delivery) error {
	body := encodeAck(ackFrame{Topic: d.Topic, Group: d.Group, Partition: d.Partition, Tags: []DeliveryTag{d.DeliveryTag}})
	return e.sendWithID(opAck, d.reqID, body)
}

// Nack returns a delivery unprocessed, optionally requeuing it (notBefore delays
// the requeue). Fire-and-forget.
func (e *Engine) Nack(d Delivery, requeue bool, notBefore *uint64) error {
	body := encodeNack(nackFrame{
		Topic: d.Topic, Group: d.Group, Partition: d.Partition,
		Tags: []DeliveryTag{d.DeliveryTag}, Requeue: requeue, NotBefore: notBefore,
	})
	return e.sendWithID(opNack, d.reqID, body)
}

func (e *Engine) sendWithID(op op, id uint64, body []byte) error {
	select {
	case e.cmdCh <- command{op: op, body: body, id: id}:
		return nil
	case <-e.done:
		return e.err()
	}
}

// ---- run-goroutine handlers --------------------------------------------

func (e *Engine) handleSubscribeOk(f frame) {
	w, ok := e.waiters[f.RequestID]
	if !ok {
		return
	}
	delete(e.waiters, f.RequestID)
	ok2, err := decodeSubscribeOk(f.Payload)
	if err != nil {
		if w.subReply != nil {
			w.subReply <- subResult{err: err}
		}
		return
	}
	prefetch := int(ok2.Prefetch)
	if prefetch < 1 {
		prefetch = 1
	}
	ch := make(chan Delivery, prefetch)
	reason := newCloseReasonCell()
	// A plain subscribe is remembered so a reconnect can restore it on the same
	// channel. The registry then owns the channel across the engine's death.
	preserve := false
	if reg := e.opts.ReconcileRegistry; reg != nil && !w.noReconcile && w.sub != nil {
		reg.register(reconcileSubscription{
			SubID:          ok2.SubID,
			Topic:          ok2.Topic,
			Partition:      ok2.Partition,
			Group:          ok2.Group,
			AutoAck:        w.autoAck,
			Prefetch:       ok2.Prefetch,
			ConsumerGroup:  w.sub.ConsumerGroup,
			ConsumerTarget: w.sub.ConsumerTarget,
			MemberID:       ok2.MemberID,
		}, ch, w.autoAck, reason)
		preserve = true
	}
	e.subs[ok2.SubID] = &subState{ch: ch, autoAck: w.autoAck, preserve: preserve, reason: reason}
	if w.subReply != nil {
		w.subReply <- subResult{sub: &Subscription{
			SubID:      ok2.SubID,
			Topic:      ok2.Topic,
			Partition:  ok2.Partition,
			Group:      ok2.Group,
			MemberID:   ok2.MemberID,
			Deliveries: ch,
			reason:     reason,
			engine:     e,
		}}
	}
}

func (e *Engine) handleDeliver(f frame) {
	d, err := decodeDeliver(f.Payload)
	if err != nil {
		return
	}
	s, ok := e.subs[d.SubID]
	if !ok {
		return // delivery for a sub we no longer track
	}
	deliv := Delivery{
		Topic:           d.Topic,
		Group:           d.Group,
		Partition:       d.Partition,
		Payload:         d.Payload,
		ContentType:     d.ContentType,
		Headers:         d.Headers,
		DeliveryTag:     d.DeliveryTag,
		Offset:          d.Offset,
		Published:       d.Published,
		PublishReceived: d.PublishReceived,
		SubID:           d.SubID,
		AutoAck:         s.autoAck,
		reqID:           f.RequestID,
		// This engine's incarnation, captured now so a later settle routes to the
		// current engine (or goes stale) without depending on this engine.
		settle:      e.settle,
		incarnation: e.incarnation,
	}
	// The channel is prefetch-sized, so this send has room under normal flow.
	// The stop guard keeps a full channel from wedging shutdown.
	select {
	case s.ch <- deliv:
	case <-e.stop:
	}
}

File: clients/csharp/Fibril/Delivery.cs
using System.Runtime.CompilerServices;
using System.Threading.Channels;

namespace Fibril;

// The delivery path: subscribe, receive pushed messages as an IAsyncEnumerable
// per subscription, and settle them with ack/nack. The actor owns the sub-id ->
// channel map and pushes deliveries. Settle is fire-and-forget and stays
// cancellation-token-free, matching the reference client's complete/retry.

/// <summary>One message pushed to a subscription.</summary>
public sealed class Delivery
{
    public string Topic { get; }
    public string? Group { get; }
    public uint Partition { get; }
    public byte[] Payload { get; }
    public ContentType ContentType { get; }
    public IReadOnlyDictionary<string, string> Headers { get; }
    public ulong Offset { get; }
    public ulong Published { get; }
    public ulong PublishReceived { get; }

    /// <summary>
    /// True when the broker already settled this delivery server-side (an auto-ack
    /// subscription). Ack/Nack are then unnecessary.
    /// </summary>
    public bool AutoAck { get; }

    internal DeliveryTag Tag { get; }
    internal ulong SubId { get; }
    private readonly ulong _reqId;   // the DELIVER frame id, reused when settling
    private readonly Engine _engine; // the connection this arrived on

    internal Delivery(Deliver d, ulong reqId, bool autoAck, Engine engine)
    {
        Topic = d.Topic;
        Group = d.Group;
        Partition = d.Partition;
        Payload = d.Payload;
        ContentType = d.ContentType;
        Headers = d.Headers ?? new Dictionary<string, string>();
        Offset = d.Offset;
        Published = d.Published;
        PublishReceived = d.PublishReceived;
        Tag = d.DeliveryTag;
        SubId = d.SubId;
        AutoAck = autoAck;
        _reqId = reqId;
        _engine = engine;
    }

    /// <summary>The delivery payload as a UTF-8 string.</summary>
    public string Text() => System.Text.Encoding.UTF8.GetString(Payload);

    /// <summary>Decodes the delivery payload from JSON into <typeparamref name="T"/>.</summary>
    public T? Json<T>()
    {
        try
        {
            return System.Text.Json.JsonSerializer.Deserialize<T>(Payload);
        }
        catch (System.Text.Json.JsonException ex)
        {
            throw new DeserializationException("json decode: " + ex.Message);
        }
    }

    /// <summary>Settles this delivery as processed, on the connection it arrived on.</summary>
    public void Ack() => _engine.Ack(this);

    /// <summary>Returns this delivery unprocessed, optionally requeuing it.</summary>
    public void Nack(bool requeue, ulong? notBefore = null) => _engine.Nack(this, requeue, notBefore);

    /// <summary>Requeues this delivery immediately for redelivery.</summary>
    public void Retry() => _engine.Nack(this, true, null);

    /// <summary>
    /// Settles this delivery as a terminal failure: not requeued, so it is
    /// dead-lettered or dropped per the queue's policy.
    /// </summary>
    public void Fail() => _engine.Nack(this, false, null);

    /// <summary>Requeues this delivery for redelivery no sooner than <paramref name="delay"/> from now.</summary>
    public void RetryAfter(TimeSpan delay)
    {
        var notBefore = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + (long)delay.TotalMilliseconds;
        if (notBefore < 0)
        {
            notBefore = 0;
        }
        _engine.Nack(this, true, (ulong)notBefore);
    }
}

/// <summary>
/// A live single-partition subscription. <see cref="Deliveries"/> yields messages
/// until the connection closes. Settle each with its Ack/Nack.
/// </summary>
public sealed class Subscription
{
    public ulong SubId { get; }
    public string Topic { get; }
    public uint Partition { get; }
    public string? Group { get; }
    internal Uuid? MemberId { get; }

    private readonly ChannelReader<Delivery> _reader;

    internal Subscription(SubscribeOk ok, ChannelReader<Delivery> reader)
    {
        SubId = ok.SubId;
        Topic = ok.Topic;
        Partition = ok.Partition;
        Group = ok.Group;
        MemberId = ok.MemberId;
        _reader = reader;
    }

    /// <summary>The stream of deliveries, ending when the connection closes.</summary>
    public IAsyncEnumerable<Delivery> Deliveries(CancellationToken ct = default) => _reader.ReadAllAsync(ct);
}

internal sealed partial class Engine
{
    // ---- publish ----

    internal void PublishUnconfirmed(PublishFrame p, CancellationToken ct)
        => Send(Op.Publish, WireOps.EncodePublish(p with { RequireConfirm = false }), ct);

    internal async Task<ulong> PublishConfirmedAsync(PublishFrame p, CancellationToken ct)
    {
        var f = await RequestAsync(Op.Publish, WireOps.EncodePublish(p with { RequireConfirm = true }), ct).ConfigureAwait(false);
        return WireOps.DecodePublishOk(f.Payload).Offset;
    }

    internal void PublishDelayedUnconfirmed(PublishDelayedFrame p, CancellationToken ct)
        => Send(Op.PublishDelayed, WireOps.EncodePublishDelayed(p with { RequireConfirm = false }), ct);

    internal async Task<ulong> PublishDelayedConfirmedAsync(PublishDelayedFrame p, CancellationToken ct)
    {
        var f = await RequestAsync(Op.PublishDelayed, WireOps.EncodePublishDelayed(p with { RequireConfirm = true }), ct).ConfigureAwait(false);
        return WireOps.DecodePublishOk(f.Payload).Offset;
    }

    // Sends a confirm-required op and returns a handle that resolves to the offset
    // later, without blocking on the confirm. The frame is written before the handle
    // returns, so callers can fire several publishes (in send order) and collect
    // their confirmations afterward to pipeline them.
    internal PublishConfirmation Confirmation(Op op, byte[] body, CancellationToken ct)
    {
        ct.ThrowIfCancellationRequested();
        var tcs = new TaskCompletionSource<Frame>(TaskCreationOptions.RunContinuationsAsynchronously);
        if (!_inbox.Writer.TryWrite(new OutgoingCommand(op, body, 0, tcs, null, false)))
        {
            throw Error();
        }

        async Task<ulong> Await()
        {
            var f = await tcs.Task.ConfigureAwait(false);
            return WireOps.DecodePublishOk(f.Payload).Offset;
        }

        return new PublishConfirmation(Await());
    }

    internal PublishConfirmation PublishWithConfirmation(PublishFrame p, CancellationToken ct)
        => Confirmation(Op.Publish, WireOps.EncodePublish(p with { RequireConfirm = true }), ct);

    internal PublishConfirmation PublishDelayedWithConfirmation(PublishDelayedFrame p, CancellationToken ct)
        => Confirmation(Op.PublishDelayed, WireOps.EncodePublishDelayed(p with { RequireConfirm = true }), ct);

    // ---- declare / topology ----

    internal async Task<DeclareQueueOk> DeclareQueueAsync(DeclareQueueFrame d, CancellationToken ct)
    {
        var f = await RequestAsync(Op.DeclareQueue, WireOps.EncodeDeclareQueue(d), ct).ConfigureAwait(false);
        return WireOps.DecodeDeclareQueueOk(f.Payload);
    }

    internal async Task<DeclarePlexusOk> DeclarePlexusAsync(DeclarePlexusFrame d, CancellationToken ct)
    {
        var f = await RequestAsync(Op.DeclarePlexus, WireOps.EncodeDeclarePlexus(d), ct).ConfigureAwait(false);
        return WireOps.DecodeDeclarePlexusOk(f.Payload);
    }

    internal async Task<TopologyOk> FetchTopologyAsync(TopologyRequest req, CancellationToken ct)
    {
        var f = await RequestAsync(Op.Topology, WireOps.EncodeTopologyRequest(req), ct).ConfigureAwait(false);
        return WireOps.DecodeTopologyOk(f.Payload);
    }

    // ---- subscribe ----

    // A plain queue subscribe is remembered for reconnect reconcile; a supervised one
    // (noReconcile) recovers via its supervisor instead. Streams resume by durable
    // cursor, so they carry no reconcile metadata.
    internal Task<Subscription> SubscribeAsync(SubscribeFrame req, bool noReconcile, CancellationToken ct)
        => SubscribeInternalAsync(Op.Subscribe, WireOps.EncodeSubscribe(req), req.AutoAck, new SubscribeMeta(req, noReconcile), ct);

    internal Task<Subscription> SubscribeStreamAsync(SubscribeStream req, CancellationToken ct)
        => SubscribeInternalAsync(Op.SubscribeStream, WireOps.EncodeSubscribeStream(req), req.AutoAck, null, ct);

    private async Task<Subscription> SubscribeInternalAsync(Op op, byte[] body, bool autoAck, SubscribeMeta? meta, CancellationToken ct)
    {
        var tcs = new TaskCompletionSource<Subscription>(TaskCreationOptions.RunContinuationsAsynchronously);
        if (!_inbox.Writer.TryWrite(new OutgoingCommand(op, body, 0, null, tcs, autoAck, meta)))
        {
            throw Error();
        }
        return await tcs.Task.WaitAsync(ct).ConfigureAwait(false);
    }

    // ---- settle ----

    internal void Ack(Delivery d)
    {
        var body = WireOps.EncodeAck(new AckFrame { Topic = d.Topic, Group = d.Group, Partition = d.Partition, Tags = new[] { d.Tag } });
        SendWithId(Op.Ack, DeliveryReqId(d), body);
    }

    internal void Nack(Delivery d, bool requeue, ulong? notBefore)
    {
        var body = WireOps.EncodeNack(new NackFrame
        {
            Topic = d.Topic,
            Group = d.Group,
            Partition = d.Partition,
            Tags = new[] { d.Tag },
            Requeue = requeue,
            NotBefore = notBefore,
        });
        SendWithId(Op.Nack, DeliveryReqId(d), body);
    }

    private static ulong DeliveryReqId(Delivery d) => DeliveryReqIds.GetValue(d, _ => new StrongBox<ulong>()).Value;

    // The DELIVER frame id travels with the Delivery so settle reuses it. It is kept
    // in a side table so Delivery stays a clean public type without an internal id field.
    private static readonly ConditionalWeakTable<Delivery, StrongBox<ulong>> DeliveryReqIds = new();

    // ---- actor-side handlers ----

    private void HandleSubscribeOk(Frame f)
    {
        if (!_waiters.Remove(f.RequestId, out var w))
        {
            return;
        }
        SubscribeOk ok;
        try
        {
            ok = WireOps.DecodeSubscribeOk(f.Payload);
        }
        catch (WireException ex)
        {
            w.SubReply?.TrySetException(ex);
            return;
        }
        var channel = Channel.CreateUnbounded<Delivery>(new UnboundedChannelOptions { SingleReader = true, SingleWriter = true });

        // A plain subscribe is remembered so a reconnect can restore it on the same
        // channel. The registry then owns the channel across the engine's death.
        var preserve = false;
        if (_opts.Reconcile is { } registry && w.Sub is { NoReconcile: false } meta)
        {
            registry.Register(new ReconcileSubscription
            {
                SubId = ok.SubId,
                Topic = ok.Topic,
                Partition = ok.Partition,
                Group = ok.Group,
                AutoAck = w.AutoAck,
                Prefetch = ok.Prefetch,
                ConsumerGroup = meta.Request.ConsumerGroup,
                ConsumerTarget = meta.Request.ConsumerTarget,
                MemberId = ok.MemberId,
            }, channel, w.AutoAck);
            preserve = true;
        }

        _subs[ok.SubId] = new SubState { Channel = channel, AutoAck = w.AutoAck, Preserve = preserve };
        w.SubReply?.TrySetResult(new Subscription(ok, channel.Reader));
    }

    private void HandleDeliver(Frame f)
    {
        Deliver d;
        try
        {
            d = WireOps.DecodeDeliver(f.Payload);
        }
        catch (WireException)
        {
            return;
        }
        if (!_subs.TryGetValue(d.SubId, out var s))
        {
            return; // delivery for a sub we no longer track
        }
        var delivery = new Delivery(d, f.RequestId, s.AutoAck, this);
        DeliveryReqIds.GetValue(delivery, _ => new StrongBox<ulong>()).Value = f.RequestId;
        s.Channel.Writer.TryWrite(delivery);
    }
}

/// <summary>
/// A handle to a confirmed publish that has been sent but not yet awaited. Call
/// <see cref="Confirmed"/> to get the broker-assigned offset. Because the frame is
/// written before the handle returns, callers can fire several publishes (in send
/// order) and collect their confirmations afterward to pipeline them.
/// </summary>
public sealed class PublishConfirmation
{
    private readonly Task<ulong> _offset;

    internal PublishConfirmation(Task<ulong> offset)
    {
        _offset = offset;
    }

    /// <summary>Waits for the broker-assigned offset of this publish.</summary>
    public Task<ulong> Confirmed(CancellationToken ct = default) => _offset.WaitAsync(ct);
}
